### PR TITLE
Add tests for fluentd validation

### PR DESCRIPTION
**Pull request by niroyb**

### CODE REVIEW

File: DCOS/fluentd-testing/fluentd.Tests.ps1 [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        Get-ServiceConfPath | Should -Not -Be $null
+    }
+}
+
+
+Describe "Fluentd logging" {
+
+    It "Can modify conf file" {
+        Copy-Item $newConfigPath $(Get-ServiceConfPath) -force
+        $LASTEXITCODE | Should -Be 0
```
Comment (ionutbalutoiu) [inline DCOS/fluentd-testing/fluentd.Tests.ps1]: `Copy-Item` from above is a PowerShell cmdlet and it won't set the `$LASTEXITCODE` as it throws exceptions on errors. So this check can be safely removed.

I would say the correct check is:

```
{ Copy-Item $newConfigPath $(Get-ServiceConfPath) -force } | Should Not Throw
```

This is an [example](https://github.com/cloudbase/juju-powershell-modules/blob/9a5af4f1e1be182daae2739ef2fc090fa274a3f6/JujuHelper/Tests/JujuHelper.Tests.ps1#L41) of an assertion of this type.

File: DCOS/fluentd-testing/fluentd.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    It "Can see changes to file in dynamic directory" {
+        [String] $DateStamp = get-date -Format yyyyMMddTHHmmss
+        $folderpath = "$env:SystemDrive/DCOS/mesos/$DateStamp"
+        $folder = md $folderpath
+        "Hello at $DateStamp" | Out-File -FilePath "$folderpath/stdout" -Encoding ascii
+
+        # Leave time for fluentd to detect change
+        Sleep 2
```
Comment (ionutbalutoiu) [inline DCOS/fluentd-testing/fluentd.Tests.ps1]: Are `Sleep 2` enough for fluentd to detect changes ?

I would recommend that we poll the fluentd output file for about 30 seconds and if there's any change there, we consider it a failure.

It's safer than sleeping 2 seconds and then see if the change is propagated.

File: DCOS/fluentd-testing/fluentd.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        $folderpath = "$env:SystemDrive/DCOS/mesos/$DateStamp"
+        $folder = md $folderpath
+        "Hello at $DateStamp" | Out-File -FilePath "$folderpath/stdout" -Encoding ascii
+
+        # Leave time for fluentd to detect change
+        Sleep 2 
+
+        $strings = Select-String -Path "$env:SystemDrive/DCOS/*.log" -Pattern "$DateStamp"
+        $strings | Should -Not -Be $null
+        # Remove-Item $folder -Recurse -Force
```
Comment (ionutbalutoiu) [inline DCOS/fluentd-testing/fluentd.Tests.ps1]: Remove commented line if not needed anymore.

File: DCOS/fluentd-testing/preprovision-agent-windows.ps1 [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    
+    $process = Start-Process "msiexec.exe" -ArgumentList $MSIArguments -Wait -NoNewWindow -PassThru
+    if ($process.ExitCode -ne 0) {
+        throw "ERROR: Failed to install Fluentd. Error code: " + $process.ExitCode + " See details in log: $logfile"
+    }
+}
+
+function Register-FluentdService () {
+    Write-Host "Registering fluentd service..."
+    fluentd --reg-winsvc i
```
Comment (ionutbalutoiu) [inline DCOS/fluentd-testing/preprovision-agent-windows.ps1]: Is `fluentd` added to the `$env:PATH` when executing this code ?
Also, this is a call to a binary from the disk, so we need to check `$LASTEXITCODE` for any non-zero exit code.

After `fluentd` call, please add this check:

```
if($LASTEXITCODE) {
    Throw "Failed to register fluentd Windows service"
}
```

File: DCOS/fluentd-testing/preprovision-agent-windows.ps1 [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    fluentd --reg-winsvc-fluentdopt "-c $env:SystemDrive/opt/td-agent/etc/td-agent/td-agent.conf -o $env:SystemDrive/opt/td-agent/td-agent.log"
```
Comment (ionutbalutoiu) [inline DCOS/fluentd-testing/preprovision-agent-windows.ps1]: Add a check for `$LASTEXITCODE` as described above

File: DCOS/fluentd-testing/preprovision-agent-windows.ps1 [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+function Start-FluentdService () {
+    Write-Host "Starting fluentd service..."
+    Start-Service fluentdwinsvc
+    
+    Write-Output '{"message":"hello from powershell"}' | fluent-cat debug.event
```
Comment (ionutbalutoiu) [inline DCOS/fluentd-testing/preprovision-agent-windows.ps1]: We're calling `fluent-cat` which I assume is a `exe` binary from the disk. Same question as above: Is it added to the `$env:PATH` at this point ? Also, please add a check for `$LASTEXITCODE`

File: DCOS/fluentd-testing/preprovision-agent-windows.ps1 (hunk 4 above)
Comment (niroyb) [reply DCOS/fluentd-testing/preprovision-agent-windows.ps1]: fluentd is added to the Path at the top. Added the exit code check

Comment (ionutbalutoiu) [reply DCOS/fluentd-testing/preprovision-agent-windows.ps1]: I wasn't sure if that refreshes the Path for the current session as well. But if that's the case, we're good.

**Pull request merged**